File: ps/swea/othello.py
import sys
sys.stdin = open("othello_input.txt")

up = (-1, 0)
down = (1, 0)
left = (0, -1)
right = (0, 1)
upright = (-1, 1)
downright = (1, 1)
downleft = (1, -1)
upleft = (-1, -1)
directions = [up, down, left, right, upright, downright, downleft, upleft]

T = int(input())
def getnext(center, direction):
    return tuple(map(sum, zip(center, direction)))

def getbefore(center, direction):
    return center[0] - direction[0], center[1] - direction[1]

def isvalid(N, center):
    return 0 < center[0] <= N and 0 < center[1] <= N

def magicmethod(N, center, color):
    grid[center[0]][center[1]] = color;
    for direction in directions:
        flag = 0
        next = getnext(center, direction)
        while isvalid(N, next):
            if grid[next[0]][next[1]] == 0:
                break
            elif grid[next[0]][next[1]] == color:
                flag = 1;
                break
            next = getnext(next, direction);

        if flag:
            next = getbefore(next, direction);
            while grid[next[0]][next[1]] != color:
                grid[next[0]][next[1]] = color;
                next= getbefore(next, direction);
        flag = 0;

        

for tc in range(1, T+1):
    N, M = map(int, input().split())
    grid = [[0 for _ in range(N+1)] for _ in range(N+1)]
    n = int(N/2)
    grid[n][n] = 1
    grid[n+1][n+1] = 1
    grid[n+1][n] = 2
    grid[n][n+1] = 2

    for _ in range(M):
        c, r, color = map(int, input().split())
        center = (r, c)
        magicmethod(N, center, color)
    countwhite = 0
    countblack = 0
    for x in range(1, N+1):
        for y in range(1, N+1):
            if grid[x][y] == 1: countblack +=1
            elif grid[x][y] == 2: countwhite += 1
    print(f"#{tc}",countblack, countwhite)

File: ps/swea/othello.cpp
#include <iostream>
#include <cstdio>
#include <vector>
#include <string>
#include <fstream>

using namespace std;
// up right down left upright, downright, downleft, upleft,
int dr[8] = {-1, 0, 1, 0, -1, 1, 1, -1};
int dc[8] = {0, 1, 0, -1, 1, 1, -1, -1};

bool isvalid(int N, int r, int c) { 
    return 1 <= c && c <= N && 1 <= r && r <= N;
}

void othello(int grid[][9], int N, int c, int r, int team) {
    int j;
    grid[r][c] = team;
    for (int i = 0; i < 8; i++) {
        j = 1; 
        while (j <= N && isvalid(N, r+dr[i]*j, c+dc[i]*j) && grid[r+dr[i]*j][c+dc[i]*j] != 0 && grid[r+dr[i]*j][c+dc[i*j]] !=team) {
        j++;
        }
        if (isvalid(N, r+dr[i]*j, c+dc[i]*j) && grid[r+dr[i]*j][c+dc[i]*j] == team) {
            for (int temp = j; j > 0; j--) {
                grid[r+dr[i]*temp][c+dc[i]*temp] =team; 
            }
        } 
}
}

int main(void) {
    ifstream input;
    input.open("othello_input.txt");

    int T;
    int c, r, team, N, M;
    int grid[9][9];
    input >> T;
    int black, white;
    
    for (int i = 0; i < T; i++) {
        input >> N >> M;
        for (int a = 1; a <= 8; a++) {
            for(int b = 1; b <= 8; b++) {
                grid[a][b] = 0;
            }
        }
        int a = N/2;
        grid[a][a] = 1;
        grid[a][a+1] = 2;
        grid[a+1][a+1] = 1;
        grid[a+1][a] = 2;

        for (int j = 0; j < M; j++) {
            input >> c >> r >> team;
            grid[r][c] = team;
            othello(grid, N, c, r, team); 
        }
        black = 0;
        white = 0;

        for (int s = 1; s <= N; s++) {
            for (int q =1; q <= N; q++) {
                if (grid[s][q] == 1) white++;
                if (grid[s][q] == 2) black++;
            }
        }
        cout << black << white;

        
    }


    return 0;
}
